src/timely-util/src/reclock.rs

```diff
@@ -0,0 +1,813 @@
+                    match remap.peek() {
+                        Some((_, t, _)) => cur_time = t.clone(),
+                        None => break,
+                    }
+                }
+
+                // STEP 5. Downgrade capability and compact remap trace
+                capset.downgrade(&reclocked_source_frontier.borrow());
+                remap_since = reclocked_source_frontier;
+                for (_, t, _) in remap_trace.iter_mut() {
+                    t.advance_by(remap_since.borrow());
+                }
+                consolidation::consolidate_updates(&mut remap_trace);
+                remap_trace
+                    .sort_unstable_by(|(_, t1, _): &(_, IntoTime, _), (_, t2, _)| t1.cmp(t2));
+            }
+
+            // STEP 6. Tidy up deferred updates
+            //         Deferred updates are represented as a list chain batches where each batch
+            //         contains two times the updates of the batch proceeding it. This organization
+            //         leads to a logarithmic number of batches with respect to the outstanding
+            //         number of updates.
+            deferred_source_updates.sort_unstable_by_key(|b| Reverse(b.len()));
+            if !new_source_updates.is_empty() {
+                deferred_source_updates.push(new_source_updates);
+            }
+            loop {
+                match (deferred_source_updates.pop(), deferred_source_updates.pop()) {
+                    (Some(a), Some(b)) if a.len() >= b.len() / 2 => {
+                        deferred_source_updates.push(a.merge_with(b));
+                    }
+                    (a, b) => {
+                        deferred_source_updates.extend(b);
+                        deferred_source_updates.extend(a);
+                        break;
+                    }
+                }
+            }
+        }
+    });
+
+    (Box::new(pusher), reclocked.as_collection())
+}
+
+/// A batch of differential updates that vary over some partial order. This type maintains the data
+/// as a set of disjoint chains that allows for efficient extraction of batches given a frontier.
+#[derive(Debug)]
+struct ChainBatch<D, T, R> {
+    /// A list of chains (sets of mutually comparable times) sorted by the partial order.
+    chains: Vec<VecDeque<(D, T, R)>>,
+}
+
+impl<D, T: Timestamp, R> ChainBatch<D, T, R> {
+    /// Extracts all updates with time not greater or equal to any time in `upper`.
+    fn extract<'a>(
+        &'a mut self,
+        upper: AntichainRef<'a, T>,
+    ) -> impl Iterator<Item = (D, T, R)> + 'a {
+        self.chains.retain(|chain| !chain.is_empty());
+        self.chains.iter_mut().flat_map(move |chain| {
+            // A chain is a set of mutually comparable elements so we can binary search
+            // to find the prefix that is not beyond the frontier and extract it.
+            let idx = chain.partition_point(|(_, time, _)| !upper.less_equal(time));
```

Comment:
Go through linearly instead of binary search since we touch those elements change

Reply:
fixed